### ISSUE TITLE
Inconsistent use of "configured" vs. "unchanged" when applying configuration files

### ISSUE THREAD

**Issue by WesCossick** [kubernetes/kubernetes#66450]
**Is this a BUG REPORT or FEATURE REQUEST?**: bug report

/kind bug

**What happened**: When deploying an unchanged resource to Kubernetes using `kubectl apply`, output sometimes states `[resource kind] "example" configured`.

**What you expected to happen**: For the output to state `[resource kind] "example" unchanged` instead.

**How to reproduce it (as minimally and precisely as possible)**: I attempted to search the codebase to determine how Kubenetes or `kubectl` decided whether to output "configured" vs. "unchanged" when applying a configuration file, but was unsuccessful. With a better understanding of how that determination occurs, I could most likely provide more examples. There is one example now in the next paragraph.

**Anything else we need to know?**:
There appears to be a number of inconsistencies with how Kubernetes determines whether to output that a resource has been "configured" or is "unchanged." I'll do my best to explain all that I could find.

When applying Deployments locally to Minikube, if the configuration file is created right before applying it using `kubectl`, the Deployment is always labeled as "configured," even if the contents are identical to the previously applied file. This behavior does not occur with Services or Ingresses, though, both of which can have their configuration files created just before applying and the output will correctly state "configured" or "unchanged" based on whether the configuration has actually changed or not. We thought maybe for Deployments, the determination was based on the last modified time of the file, so we forced a static last modified time before running `kubectl apply`, but that did not fix the issue. Therefore, we don't believe that it is making the determination based on the last modified time.

However, when applying Ingresses to GKE, they become affected by this issue, even though they weren't affected when applying to Minikube. Services, on the otherhand, still aren't affected by this issue using GKE.

In addition, some resource kinds seem to always be considered "configured," even when you apply the same file back-to-back. For example, if `test.yml` contains:

```yml
apiVersion: v1
kind: Namespace
metadata:
  name: "test"
```

...and you run `kubectl apply -f test.yml` multiple times, it will always state `namespace "test" configured` even though the configuration file isn't changing in any way. This seems to be true of CustomResourceDefinitions and StorageClasses too (probably others as well).

In general, what I'd like to see is consistent use of "configured" vs. "unchanged" when applying a resource via `kubectl apply`. This functionality appears to work now for some kinds of resources, but not all kinds. You might be wondering, why is this even important? When applying a large group of resource files, it's very helpful to know which ones have actually changed and which ones haven't. It would become immediately obvious when a resource changes when it shouldn't have, and you could investigate the issue right away, without waiting for that problem to manifest itself in other ways. As it stands now, so many are labeled as "configured" that it's not useful or accurate to rely on this statement.

**Environment**:
- Kubernetes version (use `kubectl version`): `v1.10.0` (Minikube) `v1.10.4-gke.2` (GKE)
- Cloud provider or hardware configuration: GKE
- OS (e.g. from /etc/os-release): Locally, macOS 10.13.5
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by flauflauf** [comment]
This is crucial when following GitOps practices.

We manage our Kubernetes configuration in Git. Every commit triggers a CI pipeline which applies the configuration to our cluster. Seeing "configured" in the kubectl apply output for configuration files that did not change should be a sign of configuration drift (someone directly administered the cluster without a git commit).

**Comment by nikhita** [comment]
kubectl patch has been fixed: https://github.com/kubernetes/kubernetes/pull/67385

issue for kubectl apply: https://github.com/kubernetes/kubernetes/issues/52577

**Comment by ChrisCooney** [comment]
Anyone got any thoughts on this?

**Comment by dixudx** [comment]
I think @nikhita has already solved this in #67385. It should works for kubectl version >= v1.12.0.

**Comment by IRDonch** [comment]
I found another case where `kubectl apply` will always report `configured`, even if the definition is unchanged. Try this file:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test
  labels:
    app: test
spec:
  selector:
    matchLabels:
      app: test
  strategy:
    rollingUpdate: {}
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - image: busybox
        name: busybox
        args: ['sleep', '99999']
```

Running in verbose mode reveals that every time `kubectl` tries to apply the following patch:

```json
{"spec":{"strategy":{"$retainKeys":["rollingUpdate"]}}}
```

This can be worked around by explicitly adding `type: RollingUpdate` to the `strategy` map, but I would still consider it a bug.

(This is with kubectl v1.14.0, server v1.13.4.)

**Comment by javefang** [comment]
I can confirm @IRDonch 's workaround also works on kubectl v1.13.4 and server v1.13.4.

**Comment by javefang** [comment]
I also notice the unit you use in the `resources` section matters.

e.g.

This is fine (apply again will result in `unchanged`)

```
resources:
  requests:
    cpu: "500m"
    memory: "32Mi"
  limits:
    cpu: "1"
    memory: "128Mi"
```

This will show `configured` every time you apply (when using decimal to define cpu)

```
resources:
  requests:
    cpu: "0.5"
    memory: "32Mi"
  limits:
    cpu: "1"
    memory: "128Mi"
```

This is also problematic (when using "1000m" for cpu instead of "1")

```
resources:
  requests:
    cpu: "500m"
    memory: "32Mi"
  limits:
    cpu: "1000m"
    memory: "128Mi"
```

And this (when using integer 1 for cpu instead of string "1")

```
resources:
  requests:
    cpu: "500m"
    memory: "32Mi"
  limits:
    cpu: 1
    memory: "128Mi"
```

My feeling is, the units for these are always normalised according to certain rules when applying, but the `last-applied-configuration` annotation always store the original yaml as is, causing the mismatch when comparing.

**Comment by eyalzek** [comment]
This is also the case for `StatefulSet` resources that define `volumeClaimTemplates`. Even though the resource is unchanged, `kubectl` will still try to patch the claim template. Can be reproduced with the following minimal yaml:

```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: foo
spec:
  selector:
    matchLabels:
      app: foo
  serviceName: foo
  template:
    metadata:
      labels:
        app: foo
    spec:
      containers:
      - image: busybox
        name: foo
  volumeClaimTemplates:
  - metadata:
      name: foo
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi

---
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  ports:
  - name: foo
    port: 1234
  selector:
    app: foo
```
Applying for the first time:

```
$ kubectl apply -f foo.yaml 
statefulset.apps/foo created
service/foo created
```

and on any subsequent `kubectl apply`:

```
$ kubectl apply -f foo.yaml 
statefulset.apps/foo configured
service/foo unchanged
```

Increasing verbosity shows that `kubectl` attempts to submit the following patch:
`{"spec":{"volumeClaimTemplates":[{"metadata":{"name":"foo"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"1Gi"}}}}]}}`

Statefulsets without a `volumeClaimTemplates` obviously do not have this issue. I tried every combination of fields and formats for the `volumeClaimTemplates` I can imagine, but the result is always the same. Tested on minikube running version `v1.14.0` and on GKE version `1.12.5-gke.5`. `kubectl` version used is `v1.14.1`.

**Comment by tpoindessous** [comment]
I don't think this is resolved

Thanks

/remove-lifecycle stale

**Comment by kautsig** [comment]
For those suffering this issue, here is how to debug it:

1) Check if there is a difference between your resource and the applied one `kubectl diff -f deployment.yaml`. If this gives you an empty result, I would expect a "unchanged" message on apply.
2) Run `kubectl apply -f deployment.yaml -v=8`
3) Look for the `PATCH` request. The line above it should contain the `Request Body:`
4) Analyze the patch and try to change your yaml so the patch is not requested. Go to 2.

The issues we saw boil down to:
* Resource request/limits as mentioned above
* StatefulSets with volumeClaimTemplates as mentioned above
* `azureFile` volume with `readOnly: false` 

We could change you yamls to not report change, except for the volumeClaimTemplates issue.

**Comment by silviogutierrez** [comment]
My own research, applying cert-manager directly:

`kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v0.11.0/cert-manager.yaml`

That will cause a few resources to re-configure even though there are no changes.

I used @kautsig 's technique and found the culprits. The manifest above has a `annotations:` map with nothing in it. It also has `apiGroup: ""` in other resources. These I imagine get dropped when applied to the server. `kubectl get <resource> -o yaml` does not show them.

But if I run `kubectl apply --server-dry-run` or actually apply, it shows `configured`. Oddly enough, if I run `kubectl diff` it does *not* show any diffs, which means that apply/diff behave differently. Directly negating this: https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/ (last section).

I'm not sure who is at fault here. Cert manager for using these no-op values? Kubernetes server for dropping them? Or the diff/apply logic not accounting for these being false positives?

FWIW: one of the empty fields was inside a deployment template, but the other was for a `RoleBinding.subjects.apiGroup` field, so nothing inside a template.

**Comment by sapkra** [comment]
It's just a guess and I havn't tested it yet: Maybe it's the order of the fields because they are becoming ordered through the apply.

**Comment by apelisse** [comment]
> But if I run kubectl apply --server-dry-run or actually apply, it shows configured. Oddly enough, if I run kubectl diff it does not show any diffs, which means that apply/diff behave differently. Directly negating this: https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/ (last section)

The status when running `kubectl apply` is decided based on whether a patch is sent, it doesn't mean that the patch is doing anything (sometimes kubectl doesn't realize or can't know that the patch does nothing). `kubectl diff` will send the patch, but if the patch does nothing, then nothing is changed.

**Comment by zerkms** [comment]
@apelisse now it sounds like a famous philosophical experiment: https://en.wikipedia.org/wiki/If_a_tree_falls_in_a_forest

"If diff decided to send a PATCH but nobody seen that patch change anything - did it happen or not"

**Comment by apelisse** [comment]
:D if `diff` returns nothing, then nothing happened ;-)

**Comment by apelisse** [comment]
One important thing to note is that even if kubectl sends a patch that shouldn't do anything, a mutating webhook can decide to alter the object.

**Comment by spacecat** [comment]
Using Kubernetes version 1.19.3 here and I'm also experiencing the "configured" issue.

Got some resource requests and limits in my yml mentioned by others by I have them in other Deployments as well but for those yml files get "unchanged".

One obvious difference is the `env:` config for the pod spec in my Deployment for the particular file giving me the "configured" problem. Don't know if it could be related?

**Comment by Gatsby-Lee** [comment]
> I also notice the unit you use in the `resources` section matters.
> 
> e.g.
> 
> This is fine (apply again will result in `unchanged`)
> 
> ```
> resources:
>   requests:
>     cpu: "500m"
>     memory: "32Mi"
>   limits:
>     cpu: "1"
>     memory: "128Mi"
> ```
> 
> This will show `configured` every time you apply (when using decimal to define cpu)
> 
> ```
> resources:
>   requests:
>     cpu: "0.5"
>     memory: "32Mi"
>   limits:
>     cpu: "1"
>     memory: "128Mi"
> ```
> 
> This is also problematic (when using "1000m" for cpu instead of "1")
> 
> ```
> resources:
>   requests:
>     cpu: "500m"
>     memory: "32Mi"
>   limits:
>     cpu: "1000m"
>     memory: "128Mi"
> ```
> 
> And this (when using integer 1 for cpu instead of string "1")
> 
> ```
> resources:
>   requests:
>     cpu: "500m"
>     memory: "32Mi"
>   limits:
>     cpu: 1
>     memory: "128Mi"
> ```
> 
> My feeling is, the units for these are always normalised according to certain rules when applying, but the `last-applied-configuration` annotation always store the original yaml as is, causing the mismatch when comparing.

Thank you for sharing this. I was able to get some idea from your comment :)

**Comment by ThomasSteinbach** [comment]
This issue makes GitOps tools like Flux ( https://github.com/fluxcd ) and ArgoCD ( https://github.com/argoproj ) never finish their sync of resources. Or in other words - GitOps tools try to synchronize the disired state of the repository with the actual state of the cluster resources until `kubectl apply` will report `unchanged`. However for some resources - especially custom resource definitions - kubectl will always report `configured`. Example:

```
$ kubectl apply -f https://raw.githubusercontent.com/traefik/traefik-helm-chart/master/traefik/crds/ingressroute.yaml
customresourcedefinition.apiextensions.k8s.io/ingressroutes.traefik.containo.us created
$ kubectl apply -f https://raw.githubusercontent.com/traefik/traefik-helm-chart/master/traefik/crds/ingressroute.yaml
customresourcedefinition.apiextensions.k8s.io/ingressroutes.traefik.containo.us configured
$ kubectl apply -f https://raw.githubusercontent.com/traefik/traefik-helm-chart/master/traefik/crds/ingressroute.yaml
customresourcedefinition.apiextensions.k8s.io/ingressroutes.traefik.containo.us configured
...
```

Please - this need to be fixed.

**Comment by hholst80** [comment]
Can someone try and just remove the line breaks from the generated base64 encoding of the secrets and see if that will fix the problem? As normal ConfigMap works just fine I strongly suspect the base64 encoding is not all right / consistent and that is why K8s will think Secrets have changed even though they have not.

**Comment by minhpq-vbee** [comment]
I faced same issue after added namespace. Still no update/fix until now?

**Comment by alexzielenski** [comment]
#106388 should have fixed a lot of cases reported in this thread. There is a specific test added for the case of quantities getting normalized, but it should also resolve issues with SSA not detecting identical objects when webhooks are involved and other cases with defaulting.

It'd be helpful to know if some of the cases reported in this thread are resolved by that change.

Otherwise, I think it'd be best for individual cases to open up separate issues to help track each cause.

cc @apelisse

**Comment by alexzielenski** [comment]
Looking through this thread, this case discussed earlier is still occurring:
https://github.com/kubernetes/kubernetes/issues/66450#issuecomment-478000730

As noted elsewhere in this thread, this instance of the bug is cause by the `+retainKeys` annotation on `Deployment.Spec.Strategy`: subsequent applications of the same patch without a Spec.Strategy.Type field cause kubectl not to be able to detect if the patch is unchanged from the information it has.

**TLDR: SMP can't tell if the object is changed (probably due to server-side defaulting), and sends a patch to the server. Issue arises when kubectl unconditionally reports patches sent to the server as "configured"**

Some personal research on how this manifests:

Kubectl apply calls Patch which in turn calls patchSimple with the current object and the patch source:

https://github.com/kubernetes/kubernetes/blob/6d7c270cf83eb4cff82a0663028512ff045bfd63/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L254-L259

`patchSimple` chooses to use Strategic Merge Patch, and starts to assemble a `deltaMap` (list of additions to the original object according to comments)

https://github.com/kubernetes/kubernetes/blob/0ee00ba10438a9a010d10fcc4e2301dc05fe7b68/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L2070-L2081

`deltaMap` value ends up being what was shared earlier in this thread: `{"spec":{"strategy":{"$retainKeys":["rollingUpdate"]}}}` due to the following logic:

https://github.com/kubernetes/kubernetes/blob/0ee00ba10438a9a010d10fcc4e2301dc05fe7b68/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L225-L232

kubectl has no way of knowing the `type` field is going to be populated server side if it is not present (not sure where this is done, I assume during defaulting). Because of that, I think this is correct behavior for kubectl to send the patch.

What may not be correct behavior, however, is for kubectl to show to the user that the object is "configured" instead of unchanged, when the server responds with an unchanged object:

https://github.com/kubernetes/kubernetes/blob/66c2f6069ebbfa41d0bef0f3959abda59b8483f1/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L166-L168

Indeed in local experiments, if this function returns "{}" for `patch` if the `ResourceVersion` is unchanged, then kubectl reports correctly whether the object was changed or not. I've sent a PR with this change #111579 for this bug

**Comment by alexzielenski** [comment]
Another case solved by #111579 was reported in this comment above relating to StatefulSets: https://github.com/kubernetes/kubernetes/issues/66450#issuecomment-484066218

It happens for essentially the same reason. Fields populated presumably by the server are appearing in the original object which are not in the patch yaml; so kubectl always sends the patch and reports "configured".

**Comment by xakraz** [comment]
If this has been fixed, starting which version of `kubectl` should we be able to get this behavior? 😇

**Comment by alexzielenski** [comment]
unfortunately my patch was already considered and rejected because it could be susceptible to a race condition (showing you `configured` if another applier had issued a patch at the same time as you). 

apologies to get hopes up.

to fix without such a race condition would require a patch on the server-side.

**Comment by josemrs** [comment]
This is happening to me when deploying custom CRDs for Kong Gateway. It is as https://github.com/kubernetes/kubernetes/issues/66450#issuecomment-1087703908 said, ArgoCD does not finish sync so it does not progress.

Every time I run:
`kubectl apply -f https://raw.githubusercontent.com/Kong/charts/kong-2.31.0/charts/kong/crds/custom-resource-definitions.yaml
`
I get:
```
[customresourcedefinition.apiextensions.k8s.io/ingressclassparameterses.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/ingressclassparameterses.configuration.konghq.com) unchanged
[customresourcedefinition.apiextensions.k8s.io/kongclusterplugins.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/kongclusterplugins.configuration.konghq.com) unchanged
[customresourcedefinition.apiextensions.k8s.io/kongconsumergroups.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/kongconsumergroups.configuration.konghq.com) unchanged
[customresourcedefinition.apiextensions.k8s.io/kongconsumers.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/kongconsumers.configuration.konghq.com) unchanged
[customresourcedefinition.apiextensions.k8s.io/kongingresses.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/kongingresses.configuration.konghq.com) configured
[customresourcedefinition.apiextensions.k8s.io/kongplugins.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/kongplugins.configuration.konghq.com) unchanged
[customresourcedefinition.apiextensions.k8s.io/kongupstreampolicies.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/kongupstreampolicies.configuration.konghq.com) configured
[customresourcedefinition.apiextensions.k8s.io/tcpingresses.configuration.konghq.com](http://customresourcedefinition.apiextensions.k8s.io/tcpingresses.configuration.konghq.com) unchanged
[customresourcedefinition.apiextensions.k8s.io/udpingresses.configuration.konghq.com
[Kong_CRD_diff_2.31.0.log](https://github.com/kubernetes/kubernetes/files/13959784/Kong_CRD_diff_2.31.0.log)
](http://customresourcedefinition.apiextensions.k8s.io/udpingresses.configuration.konghq.com) unchanged
```
kongupstreampolicies and kongingresses keep being configured (and our ArgoCD keeps showing them as out of sync)
I traced down the last CRDs I can apply and they sticks, it's 2.30.0.

I have attached a diff after I did:
```
kubectl apply -f https://raw.githubusercontent.com/Kong/charts/kong-2.30.0/charts/kong/crds/custom-resource-definitions.yaml
kubectl diff -f https://raw.githubusercontent.com/Kong/charts/kong-2.31.0/charts/kong/crds/custom-resource-definitions.yaml
```

[Kong_CRD_diff_2.31.0.log](https://github.com/kubernetes/kubernetes/files/13959786/Kong_CRD_diff_2.31.0.log)


Any chance we can get this prioritised?

**Comment by dnviti** [comment]
At least for pull secrets you need to configure it with the multiline chomping with no newling at the end: "|-"

For Example:

```yaml
kind: Secret
apiVersion: v1
metadata:
  name: my-pull-secret
  annotations:
    helm.sh/hook: 'pre-install, pre-upgrade'
    helm.sh/hook-delete-policy: 'before-hook-creation'
    helm.sh/hook-weight: '-999'
data:
  .dockerconfigjson: |-
    YOUR-SECRET
type: kubernetes.io/dockerconfigjson
```

This because it adds a \n at the end and when confronting with the desired manifest it results in a different string so it will be treated as different and marked as "changed".

**Comment by hholst80** [comment]
So, exactly what I said...

**Comment by Hipska** [comment]
I had this issue as well, and this comment solved it: https://github.com/kubernetes/kubernetes/issues/66450#issuecomment-478000730 (`RollingUpdate` strategy)
